apic_ml2/neutron/services/l3_router/apic_driver.py

```diff
@@ -0,0 +1,246 @@
+        return atenant_id, arouter_id, anet_id, asubnet_id
+
+    def _get_port_id_for_router_interface(self, context, router_id, subnet_id):
+        filters = {'device_id': [router_id],
+                   'device_owner': [q_const.DEVICE_OWNER_ROUTER_INTF],
+                   'fixed_ips': {'subnet_id': [subnet_id]}}
+        ports = self._plugin._core_plugin.get_ports(context.elevated(),
+                                                    filters=filters)
+        return ports[0]['id']
+
+    def _add_apic_router_interface(self, context, router_id, interface_info):
+        # Update router's state first
+        router = self._plugin.get_router(context, router_id)
+        self.update_router_postcommit(context, router)
+
+        # Add router interface
+        if 'subnet_id' in interface_info:
+            subnet = self._plugin._core_plugin.get_subnet(
+                context, interface_info['subnet_id'])
+            network_id = subnet['network_id']
+            port_id = self._get_port_id_for_router_interface(
+                context, router_id, interface_info['subnet_id'])
+        else:
+            port = self._plugin._core_plugin.get_port(
+                context, interface_info['port_id'])
+            network_id = port['network_id']
+            port_id = interface_info['port_id']
+
+        network = self._plugin._core_plugin.get_network(context, network_id)
+        tenant_id = network['tenant_id']
+        if (tenant_id != router['tenant_id'] and
+                self.aci_mech_driver.per_tenant_context and
+                not self.aci_mech_driver._is_nat_enabled_on_ext_net(network)):
+            # This operation is disallowed. Can't trespass VRFs without NAT.
+            raise InterTenantRouterInterfaceNotAllowedOnPerTenantContext()
+
+        # Map openstack IDs to APIC IDs
+        atenant_id, arouter_id, anetwork_id, _ = self._map_names(
+            context, tenant_id, router, network, None)
+
+        # Program APIC
+        self.manager.add_router_interface(
+            self.aci_mech_driver._get_network_aci_tenant(network),
+            arouter_id, anetwork_id,
+            app_profile_name=self.aci_mech_driver._get_network_app_profile(
+                network))
+        self._plugin._core_plugin.update_port_status(
+            context, port_id, q_const.PORT_STATUS_ACTIVE)
+
+    def add_router_interface_postcommit(self, context, router_id,
+                                        interface_info):
+        try:
+            self._add_apic_router_interface(context, router_id,
+                                            interface_info)
+        except Exception:
+            with excutils.save_and_reraise_exception():
+                # Rollback db operation
+                super(self._plugin.__class__,
+                      self._plugin).remove_router_interface(
+                          context, router_id, interface_info)
+
+    def remove_router_interface_precommit(self, context, router_id,
```

Comment:
I copied these from the existing l3_apic.py:

https://github.com/noironetworks/apic-ml2-driver/blob/master/apic_ml2/neutron/services/l3_router/l3_apic.py#L237-L238

and

https://github.com/noironetworks/apic-ml2-driver/blob/master/apic_ml2/neutron/services/l3_router/l3_apic.py#L259-L260

 I agree, but I didn't change this as I presumed it was done this way for a reason? I'll check with Ivar.